### PR TITLE
[SYCL] Keep track of signaled fences to properly recycle associated command-lists.

### PR DESCRIPTION
**Pull request by smaslov-intel**
This fixes the leak of "command-list"s that had their associated fences not yet signaled at the time all events in that list were already completed (sporadic timing issue). In this case the fence remained non-reset and it's command-list is not returned to the command-list cache. 

The solution is to track pending fences and reset those still in-use by the time of the queue destruction.

Signed-off-by: Sergey V Maslov <sergey.v.maslov@intel.com>

### CODE REVIEW

**Comment by bso-intel** [comment]
Could you add some description of the current issue and how this PR solved it?

**Comment by smaslov-intel** [comment]
> Could you add some description of the current issue and how this PR solved it?

Sure, I've updated the description.

**Comment by bso-intel** [comment]
> This fixes the leak of "command-list"s that had their associated fences not yet signaled at the time all events in that list were already completed (sporadic timing issue).

That's interesting. I am not sure why this ordering issue occur.
Is this a bug in L0 driver?

**Comment by smaslov-intel** [comment]
> > This fixes the leak of "command-list"s that had their associated fences not yet signaled at the time all events in that list were already completed (sporadic timing issue).
> 
> That's interesting. I am not sure why this ordering issue occur.
> Is this a bug in L0 driver?

I don't think so, the commands in the command-lists are signaling their events, and the fence is signaled after all commands are completed and thus after all events are signaled. There apparently is some time between all events are signaled but the fence is not signaled yet.

@jandres742 could you please comment on this?

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 1]
```diff
@@ -521,17 +521,19 @@ pi_result _pi_context::finalize() {
...
-  ZE_CALL(zeFenceReset(this->ZeCommandListFenceMap[ZeCommandList]));
-  ZE_CALL(zeCommandListReset(ZeCommandList));
+pi_result _pi_queue::resetCommandListFenceEntry(
+    _pi_queue::command_list_fence_map_t::value_type &MapEntry,
+    bool MakeAvailable) {
+  // Fence had been signalled meaning the associated command-list completed.
+  // Reset the fence and put the command list into a cache for reuse in PI
+  // calls.
+  ZE_CALL(zeFenceReset(MapEntry.second.ZeFence));
+  ZE_CALL(zeCommandListReset(MapEntry.first));
```
Comment (bso-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:531]: So, here we call zeFenceReset first and then zeCommandListReset.
But, when we are about to release pi_queue, the command list was successfully reset, but fence is still in use (not reset)?
Did I misunderstand this situation?

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:531]: Both the fence *and* the command-list it is fencing are not reset. These 2 always come together. The purpose of the fence is to tell if command-list is still in use.

**Comment by jandres742** [comment]
> the commands in the command-lists are signaling their events, and the fence is signaled after all commands are completed and thus after all events are signaled. There apparently is some time between all events are signaled but the fence is not signaled yet.

Correct in the sense that a fence is signaled when all commands in the list have completed, which means when all the events in the list have been signaled. If the problem is that there's could be a delay between the time all events are signaled, and when the fence is signaled, that could be true, but that would be a HW delay, as we only program the signal in the fence, we dont actually signal it.

Now, if the problem is that we need this to be able to recycle the list, then the fence is something more appropriate, as there's no guarantee that all lists have events, and an event is a per command synchronization primitive, whereas a fence is a per  list one, so it provides better guarantees that the list is ready for reuse.

Let me know if I'm following the problem statement correctly here.

**Comment by smaslov-intel** [comment]
> a delay between the time all events are signaled, and when the fence is signaled, that could be true, but that would be a HW delay

Thanks, that's exactly how I understood the issue. Now the problem with this was that we were polling the fence at events' completion, which sometime (due to mentioned delay) may be a little bit too early. In such case the fence is now remembered to be still in use, and is only polled and released (and command-list recycled) at the next earliest convenience (when a new command-list is needed, or when queue is being finalized).

**Comment by bso-intel** [comment]
Thanks, @jandres742  for clarifying the delay between the command-list and fence.
> Now the problem with this was that we were polling the fence at events' completion, which sometime (due to mentioned delay) may be a little bit too early.
Since we know there could be a delay, do we still want this checking for the fence state at event completion?

**Comment by smaslov-intel** [comment]
> Since we know there could be a delay, do we still want this checking for the fence state at event completion?

Absolutely, since this is 1) cheap and 2) gets the command-list back to the cache much earlier most of the times.

**Comment by bso-intel** [comment]
> > Since we know there could be a delay, do we still want this checking for the fence state at event completion?
> 
> Absolutely, since this is 1) cheap and 2) gets the command-list back to the cache much earlier most of the times.

I understand that with your PR the memory leak due to the timing is fixed now.
I was just asking if there is a way to delay the checking for the fence state at a little later time.
It sounds like there is no such way until we really reuse the command-list or destroy the queue.
Thanks.

**Review by bso-intel — DISMISSED** [review]
LGTM.
Your analysis was really a good for this hard problem.
NIT: It would be great to leave a comment about the potential timing issue between command-list reset and fence and that's why we need to keep the live fence list and reclaim at a later time. 
Thanks.

**Comment by bader** [comment]
> LGTM.
> Your analysis was really a good for this hard problem.
> NIT: It would be great to leave a comment about the potential timing issue between command-list reset and fence and that's why we need to keep the live fence list and reclaim at a later time.
> Thanks.

@smaslov-intel, are you going to apply this suggestion before we merge this PR?

**Comment by smaslov-intel** [comment]
Added the source code comment clarifying why we need the change.

**Review by bso-intel — APPROVED** [review]
LGTM.

**Pull request merged**